### DISCUSSION TITLE
How much can I tweak my variational parameters?

### DISCUSSION THREAD

**Discussion by FabianBallar7** [QuantumBFS/Yao.jl#512] [Q&A] [answered]
Hey all!

First, I want to express my gratitude to all the contributors to this repo. Your work has been incredibly helpful to me.

I have a question and thought this would be the best place to ask. If not, I apologize and am happy to move it elsewhere.

Here's my question:

Is there a way to constrain multiple parameters to be identical in Yao? For instance, if I have a layer where I apply several gates that all depend on a rotation by some angle $\phi$, Yao's default behavior would treat these as independent parameters. Is there a way to constrain the system so that these parameters are always equal, effectively considering them as a single variational parameter?

One way I thought of was defining the effective layer $U(\phi) = \prod_i U_i(\phi)$, however, if one uses the standard preexisting gates to do this all the parameters are still defined separately. 

Another application where this behavior could be beneficial is for example in trying to find the maximum Trotter step that minimizes the energy difference. In that case, one could write the time evolution product of the individual Trotter terms and have a unique time step in each exponential. How it's currently set up all time steps would vary independently.

If anyone has experience with this or can point me in the right direction, I would greatly appreciate it!

Additionally, if this feature isn't currently available but could be implemented, I would be happy to help contribute to its development.

Thank you in advance for your help!

**Comment by GiggleLiu** [comment, marked as answer]
Parameters sharing is supposed to be handled by an external automatic differentiation engine, e.g. [Zygote](https://github.com/FluxML/Zygote.jl).
If you do not want to port Yao to an AD engine, you can sum up the gradients for the same variable manually.

**Reply by FabianBallar7** [reply]
Thanks a lot for your reply, it cleared things up. 

Maybe just in case other people stumble onto this discussion with a similar question I'll leave a link to [this code](https://github.com/QuantumBFS/QuAlgorithmZoo.jl/blob/master/examples/PortZygote/shared_parameters.jl) where a very clear example is given that helped me work it out for my own application.
